crates/sc-proof-of-time/src/gossip.rs

```diff
@@ -91,11 +124,18 @@ impl<Block: BlockT> Validator<Block> for PotGossipValidator {
     fn validate(
         &self,
         _context: &mut dyn ValidatorContext<Block>,
-        _sender: &PeerId,
+        sender: &PeerId,
         mut data: &[u8],
     ) -> ValidationResult<Block::Hash> {
         match PotProof::decode(&mut data) {
-            Ok(_) => ValidationResult::ProcessAndKeep(topic::<Block>()),
+            Ok(proof) => {
+                if let Err(err) = self.pot_state.is_candidate(*sender, &proof) {
```

Comment:
Re: https://github.com/subspace/subspace/pull/1739#discussion_r1282459021
> 1. I'd have some trace level logging here indicating the reason of rejection

Added trace log here.
> 2. It will still forward invalid messages, meaning receiver that validates and sees that message is no good will ban peer that forwarded the message, not originator; my expectation was that we do full validation here and do not let through messages that are invalid at all

I also thought of doing AES validation here. But even if validation succeeds here, when gossip eventually gets it, only bytes are delivered again. So we have to deserialize again, and either remember proofs that passed earlier here, or repeat AES verification again.

So this is a best effort kind of thing(like Bloom filter as mentioned in comments). We discard things which we know for sure we don't need (e.g) stale proofs. But we may pass proofs that might eventually fail validation later. More experiments will show the false positive percentage

Reply:
If you didn't comment, I wouldn't even notice this. I would strongly prefer that next time comments are addressed not in last and biggest commit with a bunch of other changes, but rather separately, so it is easy to review them.

> So we have to deserialize again

Deserialization with SCALE is cheap, you can consider it free, especially once we switch from vector of proofs to static array it'll be just memcpy.

> and either remember proofs that passed earlier here, or repeat AES verification again

No additional AES verification is necessary because you have two sources: locally produced proofs and pre-verified from gossip, in both cases they are already known to be valid. That was my point all along, to remove AES verification from state completely.

Reply:
> and pre-verified from gossip

Just to make sure, is it safe to assume this? I mean, whatever shows up eventually in the gossip receive stream has already been through the gossip validator.

Reply:
To the best of my knowledge yes, it'd be very odd if it was not the case. But feel free to verify with sources and tell me if I'm wrong, thankfully all the code is available for review.

Reply:
Moved AES verify to gossip validator, state manager no longer does any verification